### PR TITLE
Teacher rule update

### PR DESCRIPTION
**Pull request by epawlowska**

### CODE REVIEW

File: whylog/teacher/__init__.py [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
 from whylog.teacher.constraint_links_base import ConstraintLinksBase
+from whylog.teacher.exceptions import NotUniqueParserName
 from whylog.teacher.user_intent import UserParserIntent, UserRuleIntent
 
-# :type line: FrontInput
-TeacherParser = namedtuple('TeacherParser', ['line', 'name', 'primary_keys', 'log_type'])
+
+class TeacherParser(object):
+    """
+    :type line: FrontInput
```
Comment (andrzejgorski) [inline whylog/teacher/__init__.py]: This docs string is inopportune.

File: whylog/teacher/__init__.py [hunk 2]
```diff
@@ -91,22 +99,29 @@ def guess_patterns(self, line_id):
         Returns list of guessed patterns for a line.
         """
         pattern_matches = self.pattern_assistant.guess_pattern_matches(line_id)
-        return [pattern_match.pattern for pattern_match in pattern_matches]
+        # TODO: Consider returning dict[int, pattern_match] in order to show guessed converters too.
```
Comment (andrzejgorski) [inline whylog/teacher/__init__.py]: The dict is better option for me.

File: whylog/teacher/__init__.py [hunk 3]
```diff
@@ -91,22 +99,29 @@ def guess_patterns(self, line_id):
...
+        return [pattern_match.pattern for pattern_match in six.itervalues(pattern_matches)]
 
     def choose_guessed_pattern(self, line_id, pattern_id):
         self.pattern_assistant.update_by_guessed_pattern_match(line_id, pattern_id)
 
     def set_pattern_name(self, line_id, name):
```
Comment (andrzejgorski) [inline whylog/teacher/__init__.py:108]: I'm not sure but I thought that Leszek & Paweł on the last meeting said that we don't need that function, and the name of pattern should be generated automaticly.

File: whylog/teacher/__init__.py [hunk 4]
```diff
@@ -91,22 +99,29 @@ def guess_patterns(self, line_id):
...
-        pass
+        names_blacklist = self._get_names_blacklist()
+        if self.config.is_free_parser_name(name, names_blacklist):
+            self._parsers[line_id].name = name
+        else:
+            raise NotUniqueParserName(name)
```
Comment (andrzejgorski) [inline whylog/teacher/__init__.py]: I'm not sure which is better solution to raise exception or return it as the result of the verify method.

File: whylog/tests/tests_teacher/__init__.py [hunk 5]
```diff
@@ -95,6 +99,57 @@ def test_default_user_parser(self):
...
+    def test_setting_primary_key(self):
+        parser = self.teacher.get_rule().parsers[self.cause1_id]
+        new_primary_key_groups = [1, 2]
+        assert not new_primary_key_groups == parser.primary_key_groups
+        self.teacher.set_primary_key(self.cause1_id, new_primary_key_groups)
+        parser = self.teacher.get_rule().parsers[self.cause1_id]
+        assert new_primary_key_groups == parser.primary_key_groups
+
+    def test_setting_log_type(self):
+        new_log_type = '_very_low_likely_log_type'
```
Comment (andrzejgorski) [inline whylog/tests/tests_teacher/__init__.py]: Use the object of the LogType class from config instead just a string. If you have any reasons to not do that convince me.

File: whylog/tests/tests_teacher/__init__.py [hunk 6]
```diff
@@ -95,6 +99,57 @@ def test_default_user_parser(self):
...
+        self.teacher.set_log_type(self.effect_id, new_log_type)
+        parser = self.teacher.get_rule().parsers[self.effect_id]
+        assert new_log_type == parser.log_type_name
+
+    def test_update_pattern(self):
```
Comment (andrzejgorski) [inline whylog/tests/tests_teacher/__init__.py:138]: If this test isn't related to new code it should be in another PR.

File: whylog/tests/tests_teacher/__init__.py [hunk 7]
```diff
@@ -95,6 +99,57 @@ def test_default_user_parser(self):
...
+        new_effect_pattern = r'^([0-9]{4}-[0-9]{1,2}-[0-9]{1,2} [0-9]{1,2}:[0-9]{1,2}:[0-9]{1,2}) Error occurred in (.*) test$'
+        self.teacher.update_pattern(self.effect_id, new_effect_pattern)
+        updated_pattern = self.teacher.get_rule().parsers[self.effect_id].pattern
+        assert new_effect_pattern == updated_pattern
+
+        not_matching_pattern = new_effect_pattern + 'not_matching_part_of_regex'
+        self.assertRaises(
+            NotMatchingRegexError, self.teacher.update_pattern, self.effect_id, not_matching_pattern
```
Comment (andrzejgorski) [inline whylog/tests/tests_teacher/__init__.py]: I'm not sure how to handle exceptions like that in whylog_vim layer. I thought that teacher.verify method should inform about that.

File: whylog/teacher/__init__.py (hunk 3 above)
Comment (epawlowska) [inline whylog/teacher/__init__.py:108]: It is by default generated automatically. This method is to update name if user would like to change it.

File: whylog/tests/tests_teacher/__init__.py (hunk 6 above)
Comment (epawlowska) [inline whylog/tests/tests_teacher/__init__.py:138]: Ok, I will avoid such tests in next PRs

File: whylog/teacher/__init__.py [hunk 8]
```diff
@@ -83,38 +88,45 @@ def update_pattern(self, line_id, pattern):
         Removes constraints related with updating line
         TODO: Update related constraints rather than remove.
         """
-        self.pattern_assistant.update_by_pattern(line_id, pattern)
+
+        try:
+            self.pattern_assistant.update_by_pattern(line_id, pattern)
+        except NotMatchingPatternError:
+            return NotMatchingPattern(self._parsers[line_id].line.line_content, pattern)
```
Comment (andrzejgorski) [inline whylog/teacher/__init__.py]: I'm not sure if this method should return error.
I don't know if it is an error or warning.

File: whylog/teacher/__init__.py [hunk 9]
```diff
@@ -83,38 +88,45 @@ def update_pattern(self, line_id, pattern):
...
     def choose_guessed_pattern(self, line_id, pattern_id):
         self.pattern_assistant.update_by_guessed_pattern_match(line_id, pattern_id)
 
     def set_pattern_name(self, line_id, name):
-        pass
+        names_blacklist = self._get_names_blacklist()
+        if self.config.is_free_parser_name(name, names_blacklist):
+            self._parsers[line_id].name = name
+        else:
+            return NotUniqueParserName(name)
```
Comment (andrzejgorski) [inline whylog/teacher/__init__.py]: Same as at line 95.

File: whylog/teacher/rule_problems.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+class TeacherRuleProblem(object):
```
Comment (andrzejgorski) [inline whylog/teacher/rule_problems.py:1]: Create 2 classes warning and errors.

Comment (epawlowska) [inline whylog/teacher/rule_problems.py:1]: As we discussed earlier with Client, it can be 1 class - Problem. Being on error list it is treated as error, being on warning list it is treated as warning.

File: whylog/teacher/__init__.py (hunk 8 above)
Comment (epawlowska) [inline whylog/teacher/__init__.py]: It is warning, every updating unit method should return warning.

File: whylog/tests/tests_teacher/__init__.py [hunk 11]
```diff
@@ -144,8 +143,8 @@ def test_update_pattern(self):
         assert new_effect_pattern == updated_pattern
 
         not_matching_pattern = new_effect_pattern + 'not_matching_part_of_regex'
-        warning = self.teacher.update_pattern(self.effect_id, not_matching_pattern)
-        assert warning.pattern == not_matching_pattern
+        problems = self.teacher.update_pattern(self.effect_id, not_matching_pattern)
+        assert type(problems[0]) == NotMatchingPattern
```
Comment (ppolewicz) [inline whylog/tests/tests_teacher/__init__.py]: don't test type, test behavior

File: whylog/tests/tests_teacher/__init__.py [hunk 12]
```diff
@@ -109,9 +109,8 @@ def test_setting_parser_name(self):
         rule = self.teacher.get_rule()
         assert new_name == rule.parsers[self.effect_id].pattern_name
 
-        self.assertRaises(
-            NotUniqueParserName, self.teacher.set_pattern_name, self.cause1_id, new_name
-        )
+        problems = self.teacher.set_pattern_name(self.cause1_id, new_name)
+        assert type(problems[0]) == NotUniqueParserName
```
Comment (ppolewicz) [inline whylog/tests/tests_teacher/__init__.py]: don't test type, test behavior

``` python
assert str(problems) == "['Name is not unique, name: something']"
```
